### ISSUE TITLE
Optimization: Reduced CPU usage on bandwidth load.

### ISSUE THREAD

**Issue by Marctraider** [porech/engarde#22]
Optimization: Reduced CPU usage on bandwidth load.

Noticed the software is quite CPU bound, although it doesn't currently appear to give me any issues, it could maybe be beneficial for higher throughput on load).

**Comment by ale-rinaldi** [comment]
Yep, I'm pretty sure CPU usage and bad throughput are really related. I never had a good enough look at this aspect, since my main usage of engarde (and the reason why I wrote it) is to reliably forward a 256Kbps audio stream across the Internet... But I think it's time to seriously work on it. Thank you

**Comment by Marctraider** [comment]
I typically noticed this when I first transitioned from ARM based router to Intel NuC and my clock speed was stuck on 800Mhz due to some linux quirks. (intel_pstate vs acpi-cpufreq driver).

At first I thought it was the cumulative bandwidth with wireguard and the encryption it uses, but that doesn't seem to be the case! :)

**Comment by ale-rinaldi** [comment]
Hello,
I just performed a pure test with the simplest thing I could compose: a Go application that forwards everything it receives from one side to another. No multiple interfaces, no custom logic, no error checks, nothing. And even on this test application, the throughtput is quite limited and the CPU usage is higher than what I except.

So, the good news are that this isn't an issue with engarde itself, but with the way Golang handles UDP sockets. So I'll need to study a little on this, but I hope to be able to address it.

**Comment by Marctraider** [comment]
I see your comment in mail, not here though. 

either way, I can sometimes sustain 175+mbps which the max my fastest connection can achieve, I assume its also drawing about 80mpbs from the two other lines, so an accumulation of 250+mbit, and at that point i'm not even sure if the application or my cpu is at its limits.

Its not a severity thing for me, but any optimization is always welcome (Mostly in regard to unable to write socket / short ping spike thingy)

No sweat :)

**Comment by ale-rinaldi** [comment]
You're right, I can't see the comment here, too :/

Well, by the way I don't like it using so much CPU anyway, it's a nonsense thing for just taking UDP packets from one place and putting them in another.

For the ping issue, maybe I have a solution for it, even if I'll need your help to test it since it's not easy for me to reproduce. I'll update the relative issue as soon as I have something ready to test.

BTW, if you're finding engarde useful for your workflow, we just added a Bitcoin address for  donations in the readme :)

**Comment by Marctraider** [comment]
Do you have paypal instead? I never really dealt in bitcoins lol.

But sure, I'm more than happy to test stuff it it makes the program (and redundancy) even more efficient.

**Comment by Marctraider** [comment]
Edit: Made a bitcoin wallet, cheers!

**Comment by ale-rinaldi** [comment]
Hi @Marctraider , wow, that's great, thank you very much! Unfortunately I'm still working on your issue, expecially regarding the latency peaks. I didn't make a test build with reduce throughput because it's simply insane: it doesn't go over 1.5Mbps and that's not a viable solution at all.

I tried different approaches about it but none of them gave the expected results: I hope to be able to find a solution soon!

Thanks again :)

**Comment by Marctraider** [comment]
@ale-rinaldi Np man, it is for the appreciation for your current code :-)

Does the 'buffer overflow' issue also happen if engarde detached a specific interface faster? In that case maybe it is more easily handled with a simple check somehow with high polling rate? To at least diminish the effects. 

Seems it is now just writing to something even if it is no longer possible. No clue how this would have to be checked though.

**Comment by talalmash** [comment]
I'm able to pull 25mbit/s (my internet limit) on 720mhz ar71xx Archer C7v2, will test over lan to see max, I simply compiled engarde-client with mips flag without any effort, works like a dream! Thank you dev!
My journey for usable internet from worse to best on unstable lines :  Speedify on RPi -> OpenMPTCProuter(best of all for TCP only + aggregate) ->  borrowed Peplink -> Glorytun UDP  -> ExitLag -> Engarde (not a bond but still a tunnel over 2 connections), finally done :)

**Comment by Marctraider** [comment]
I also tried most of those, but non can properly handle link loss without major rubberbanding. And TCP is even worse for general internet usage and gaming/voip.

Engarde beats all those solutions for stable connection ;-)

**Comment by ale-rinaldi** [comment]
Hi @talalmash , I'm really happy you're finding engarde useful, thanks for your feedback :)

**Comment by Marctraider** [comment]
Any update on this? Possible to somehow adjust code without a complete rewrite?

I wonder maybe if compiling it with more cpu specific optimizations could yield a few more megabits of throughput, but I never really compiled any GoLang stuff.

**Comment by xavierleung** [comment]
I am also curious if someone figure out any optimizations by now.

**Comment by cjangrist** [comment]
Is there a set of tests that can be run offline to triage and diagnose any CPU bottlenecking or throughput issues?

**Comment by Llorx** [comment]
> I'm able to pull 25mbit/s (my internet limit) on 720mhz ar71xx Archer C7v2, will test over lan to see max, I simply compiled engarde-client with mips flag without any effort, works like a dream! Thank you dev!
> 
> My journey for usable internet from worse to best on unstable lines: Speedify on RPi -> OpenMPTCProuter(best of all for TCP only + aggregate) -> borrowed Peplink -> Glorytun UDP (bad with packet loss, even at 1% limiter) -> ExitLag -> Engarde (no aggregation), finally done :)
> 
> Edit: And yes it spikes 100% cpu, average 75 on full speed, but with no side effect, no latency increase, no speed decrease, and even LuCI webui isn't sluggish, running stress-ng only brings the speed down to 8mbit.

Hey! So Peplink is not that good?
